velox/functions/sparksql/ToPrettyString.h

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/expression/CastExpr-inl.h"
+#include "velox/functions/Udf.h"
+#include "velox/type/Conversions.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace detail {
+static const StringView kNull = "NULL";
+}
+
+/// to_pretty_string(x) -> varchar
+/// Returns pretty string for int8, int16, int32, int64, bool, Date, Varchar. It
+/// has one difference with casting value to string:
+/// 1) It prints null input as "NULL" rather than producing null output.
+template <typename TExec>
+struct ToPrettyStringFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  template <typename A>
+  void initialize(
+      const std::vector<TypePtr>& inputTypes,
+      const core::QueryConfig& /*config*/,
+      A* /*a*/) {
+    inputType_ = inputTypes[0];
+  }
+
+  template <typename TInput>
+  Status callNullable(out_type<Varchar>& result, const TInput* input) {
+    if (input) {
+      if constexpr (std::is_same_v<TInput, StringView>) {
+        result.setNoCopy(*input);
+        return Status::OK();
+      }
+      if constexpr (std::is_same_v<TInput, int32_t>) {
+        if (inputType_->isDate()) {
+          try {
+            auto output = DATE()->toString(*input);
+            result.append(output);
+          } catch (const std::exception& e) {
+            return Status::Invalid(e.what());
```

Comment:
It appears the date-to-string cast throws for runtime error,  but here we return "NULL" for both user error and runtime error. Which is aligned with Spark?
https://github.com/facebookincubator/velox/blob/ccd765d3d9e70ddc4cb222b00433b0922915b6cc/velox/expression/CastExpr.cpp#L183-L185